bpf/bpf_lxc.c

```diff
@@ -353,6 +353,9 @@ static __always_inline int ipv6_l3_from_lxc(struct __ctx_buff *ctx,
 
 	/* The packet goes to a peer not managed by this agent instance */
 #ifdef ENCAP_IFINDEX
+# ifdef ENABLE_WIREGUARD
+	if (!dst_remote_ep)
+# endif /* ENABLE_WIREGUARD */
```

Comment:
Yes, pod2pod will be encapsulated only by the Wireguard device. I don't see why it makes sense to encapsulate pod2pod twice (first via cilium_wg0, then via cilium_vxlan). Is there any good reason? Source pod src identity will be derived from IPCache on a destination node anyway.

Reply:
Made a clarification in the code comment. Also, commented only in the IPv4 section, as with commenting both we risk that one section might get stale.

Reply:
> Yes, pod2pod will be encapsulated only by the Wireguard device. I don't see why it makes sense to encapsulate pod2pod twice (first via cilium_wg0, then via cilium_vxlan). Is there any good reason? Source pod src identity will be derived from IPCache on a destination node anyway.

Ok, but then what is the rationale to support vxlan? Not sure about reason tbh, the only thing that would come to mind is if there are security policies that would only expect vxlan/geneve traffic, but not wg traffic.

Reply:
> but then what is the rationale to support vxlan?

We currently encrypt only pod2pod with Wireguard, meaning that pod2host and host2pod won't be encrypted. If your underlying network does not recognize a pod subnet, then pod2host and host2pod packets might be dropped in the direct routing mode. Using the tunneling mode prevents from that.

Reply:
To me these semantics feel unintuitive and unexpected, in the sense that pod2pod goes over wg and not via vxlan, but pod2host and host2pod go over vxlan and not wg. I mostly worry that underlying network policies could drop wg traffic if a user explicitly opted for vxlan (or geneve) so expectation from user side would be that anything non-public would have to take the vxlan/geneve tunnel. Either way, this should clearly be documented in a user guide along with rationale. If you think it makes sense, feel free to merge under the nit to make it very clear in the doc.

Reply:
ACK. Let's merge it. Meanwhile I asked advanced tunneling users whether this won't break their setups. Also, once we start supporting pod2host, host2pod and host2host over Wireguard, then the whole tunneling mode can go away.

Reply:
Also, cc @pchaigno for visibility to this thread, as he has been recently fixing various asymmetries in the datapath.

Reply:
Sgtm, thanks!

Reply:
Agree with Daniel that it seems counterintuitive (+ it will be behaving differently from IPSec). Since this is in beta, I guess it's okay as long as properly documented. If users complain, we can always change it before going out of beta.

Reply:
What is IPSec pod2pod traffic path when tunneling is on?

Reply:
I don't think this is correct. Does it use the vxlan configured ports and IPs? If the user configured vxlan for some reason e.g. clustermesh, middle boxes, firewalls, etc. we should give them vxlan or just tell them to not config vxlan. I _agree_ the vxlan + encryption headers + encryped(payload) is a bit awkward, but this seems even more broken. If we hit this limitation why not just document why vxlan is not supported so folks understand the details.

At the moment IPSec uses tunnel mode even when vxlan is enabled, but it should use transport mode which would drop the overhead of extra headers. But, in IPSec this needs to be hidden behind yet another option because its not obvious to me yet if folks rely on encrypting original IP headers for security reasons, hiding options, tokens, etc.   The enablement of transport mode, now that we have OK CI tests, should be fairly straight forward.

Reply:
> Does it use the vxlan configured ports and IPs?

Noup, it uses k8s node IPs and Wireguard UDP port. A pod2pod encrypted packet received by a destination node on a physical device is `| L2 hdr | L3 hdr: src node IP -> dst node IP | L4 UDP hdr: 51876 -> 51876 | Encrypted L3 (src pod IP -> dst pod IP) + L4 + Payload |`.

> If the user configured vxlan for some reason e.g. clustermesh, middle boxes, firewalls, etc. we should give them vxlan or just tell them to not config vxlan.

Then, if considering the above, I think that our vxlan implementation w/o any encryption is in general broken, as for the pod2host case we don't send a packet over the tunnel device.

From my experience talking to users, the main motivation for using vxlan with Cilium is that the underlying network is not aware of pod CIDR, so it can drop packets. So I think it's fine to just document the Wireguard integration behavior.